src/wasm/wasm-type.cpp

```diff
@@ -307,89 +391,37 @@ template<typename Self> struct TypeGraphWalker : TypeGraphWalkerBase<Self> {
   std::unordered_set<Type> scannedTypes;
 };
 
-// A type graph walker that only traverses the direct HeapType children of the
-// root, looking through child Types. What to do with each child is left to
-// subclasses.
-template<typename Self> struct HeapTypeChildWalker : HeapTypeGraphWalker<Self> {
-  // Override this.
-  void noteChild(HeapType* child) {}
-
+// A type graph walker that calls `noteChild` on each each direct HeapType child
+// of the root.
+template<typename Self> struct HeapTypeChildWalker : TypeGraphWalkerBase<Self> {
   void scanType(Type* type) {
     isTopLevel = false;
-    HeapTypeGraphWalker<Self>::scanType(type);
+    if (type->isRef()) {
+      this->self().noteChild(type->getHeapType());
+    } else {
+      TypeGraphWalkerBase<Self>::scanType(type);
+    }
```

Comment:
Can you give an overview of the changes in this file? I don't see an obvious reason why the encoding change necessitates changes like this, so I feel I'm missing something.

Reply:
Since heap types are now bit-packed into the reference type encoding, the `TypeGraphWalkerBase` can no longer scan the heap type children of reference types because there is no valid pointer to just the heap type. Various users of `TypeGraphWalkerBase` therefore need to be updated to do work in `scanType` for referenced heap types that they used to do in `scanHeapType`.

The fact that tuples are now the only types that need canonicalization also allows some simplifications at the end of the routine for building a rec group.

Reply:
Wait, can't `TypeGraphWalkerBase` get the valid pointer to the HeapType from the Type by looking at the right bits?

Reply:
Since the heap type is now stored directly in the representation bits of a type, a pointer to the HeapType would be the same as the pointer to the Type. But `TypeGraphWalkerBase` expects to be able to mutate the pointed-to heap heap type (kind of like `replaceCurrent` with our normal IR walkers), and that's no longer possible because it would have to somehow know how to restore the masked-off bits.

This is the same reason bitfield members are not addressable in C/C++.

Reply:
Hmm, I can't seem to to find the code in `TypeGraphWalkerBase` that mutates HeapTypes. I just see walking code. Where can I find that?

Reply:
Actually there might not have been any `TypeGraphWalkerBase` subtypes that did mutate the heap types, even previously. But the point is that the base class was prepared for such mutations, so it had to change.

Reply:
I'm still puzzled here:

> and that's no longer possible because it would have to somehow know how to restore the masked-off bits.

Can we not leave the irrelevant bits alone? That is, given a pointer to a HeapType, we can replace only the HeapType part by
```cpp
void replaceHeapTypeInPlace(HeapType* ptr, HeapType newValue) {
  auto nullabilityBit = *ptr & 2;      // tuple bit must be zero anyhow, for a ref
  *ptr = newHeapType | nullabilityBit; // same nullability, new heap type
}
```
In general I can see why we can't use a `HeapType*` in the simple way (and so we can't just do `*ptr = newValue`). But given that we know the `HeapType*` is a `Type*`, I think we can do the above?

Reply:
Yeah, if we wanted to make this utility more powerful we could do something like that. But actually we want to make it simpler and no more powerful than necessary.